### DISCUSSION TITLE
Hermeticity and Security

### DISCUSSION THREAD

**Discussion by aakarim** [cue-lang/cue#1970] [Q&A]
Hi, it's explained in several places that the core of CUE (i.e. except the scripting layer) is hermetic. My understanding of this is that there should be no side effects, especially network or disk calls, observed while evaluating the value lattice. I have also noticed that Dagger does not allow you to express secrets in CUE because there is a danger that network/disk calls could leak secrets - how can these two points coexist?

I am asking because we are developing a new CI workflow with CUE at its heart, where there are separate CUE files in a shared package that are decrypted from Vault and loaded into the value lattice in our CI system. We don't use the CUE tooling layer for this at all, it's just plain CUE files that are decrypted and then evaluated using CUE export. Is this inherently insecure?  What are the potential attack vectors here?

**Comment by antong** [comment]
As I understand it, the hermeticity of CUE is not so much about evaluation not having external side effects. It is more that the outside environment has no effect on the output when evaluating CUE. The same CUE configuration must always evaluate to the same, regardless of the computer platform running cue, operating system, language, the time of day or time zone, file system contents, external web sites or whether there is network connectivity at all, ... and so on. Same CUE in, same output. In practice this does lead to the implementation not really having side effects.

I know nothing about dagger, but I would guess the security concerns are not about the cue evaluator posting secrets in the CUE configuration to external sites. But, if you would have secrets in (or injected into) user supplied CUE configurations, then there are risks that such secrets could end up in error messages, logs etc. Totally made up example:
```
svc: passwd: {ci.#Secret & { id: "password1" }}.text
svc: passwd: int // user supplied typo
```
Then you may end up with errors in logs, alerts etc., like:
```
svc.passwd: conflicting values "hunter2" and int (mismatched types string and int)
```
